Bump Faer to latest version and increase MSRV to 1.85

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit bumps the faer package version used by qiskit to the latest release 0.22.6 which is up from 0.19.4. We had held back on upgrading the faer release because more recent versions bumped the MSRV to 1.84 when that was *very* new. Now that we're more time out from that release we can consider bumping Qiskit MSRV. To enable using the latest faer version. Additionally, the latest faer version fixes issues identified with using the older release on Windows.

This commit  opts to raise the MSRV to 1.85 instead of 1.84, as this will enable us to use Rust Edition 2024 which has some nice QoL improvements and its worth upgrading too.

Typically we would have held back on bumping the MSRV until 3.0, but it might be some time before 3.0 is branched and we explicitly do call out in the version strategy that the minimum supported Rust version isn't part of the backwards compatibility guarantees around Qiskit [1], so we are able to bump it if there is a need to. The need to fix the issues when running on Windows also requires us to upgrade our version of faer now instead of later.



### Details and comments

TODO:

- [x] Confirm this fixes the issues reported in #14751 

[1] https://quantum.cloud.ibm.com/docs/en/open-source/qiskit-sdk-version-strategy

mtreinish committed:
Bump Faer to latest version and increase MSRV to 1.85

This commit bumps the faer package version used by qiskit to the latest
release 0.22.6 which is up from 0.19.4. We had held back on upgrading
the faer release because more recent versions bumped the MSRV to 1.84
when that was *very* new. Now that we're more time out from that release
we can consider bumping Qiskit MSRV. To enable using the latest faer
version. Additionally, the latest faer version fixes issues identified
with using the older release on Windows.

This commit  opts to raise the MSRV to 1.85 instead of 1.84,
as this will enable us to use Rust Edition 2024 which has some nice
QoL improvements and its worth upgrading too.

Typically we would have held back on bumping the MSRV until 3.0, but it
might be some time before 3.0 is branched and we explicitly do call out
in the version strategy that the minimum supported Rust version isn't
part of the backwards compatibility guarantees around Qiskit [1], so we
are able to bump it if there is a need to. The need to fix the issues
when running on Windows also requires us to upgrade our version of faer
now instead of later.

[1] https://quantum.cloud.ibm.com/docs/en/open-source/qiskit-sdk-version-strategy

mtreinish commented:
It looks like the failure is caused by https://github.com/sarah-quinones/gemm/issues/35 which is a blocker for this since we have tier 1 support for x86_64 macOS.

mtreinish committed:
Downgrade to faer 0.21

Faer 0.22 has bugs that prevent it from building on x86_64 macOS which
prevents us from using it. However the 0.21 release doesn't seem to have
these issues which should also hopefully fix the issue on windows. This
commit downgrades to the 0.21 release instead of 0.22 to see if it works
on both platforms.

raynelfss commented:
Can confirm that, for #14751, this does fix all the test failures on my local Windows runner. The branch [msrv-bump-test](https://github.com/raynelfss/qiskit/tree/msrv-bump-test) contains a rebased version of #14751 if anyone else wants to run the tests locally.

Output
```cmd
UpdateCTestConfiguration  from :C:/Users/raynelfss/Desktop/Qiskit/qiskit/qiskit/test/c/build/DartConfiguration.tcl
Test project C:/Users/raynelfss/Desktop/Qiskit/qiskit/qiskit/test/c/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
...
test 3
    Start 3: test_consolidate_blocks

3: Test command: C:\Users\raynelfss\Desktop\Qiskit\qiskit\qiskit\test\c\build\test\c\Debug\test_driver.exe "test_consolidate_blocks"
3: Working Directory: C:/Users/raynelfss/Desktop/Qiskit/qiskit/qiskit/test/c/build/test/c
3: Environment variables:
3:  PATH=%PATH%;C:/Users/raynelfss/Desktop/Qiskit/qiskit/qiskit/target/debug;C:/Users/raynelfss/AppData/Local/Programs/Python/Python312
3: Test timeout computed to be: 9999879
3: --- test_consolidate_small_block  : Ok
3: --- test_wire_order               : Ok
3: --- test_3q_blocks                : Ok
3: === Number of failed subtests: 0
3/9 Test #3: test_consolidate_blocks ..........   Passed    0.01 sec
...
100% tests passed, 0 tests failed out of 9
```

mtreinish committed:
Try an earlier nightly for miri

mtreinish committed:
Lower tolerance for quantum info tests under miri

Miri intentionally has reduced floating point precision on newer
versions of the rust compiler. Since we have to bump the nightly version
we use for the miri tests because the version we were previously using
is older than the new MSRV. This is causing some quantum info tests to
fail where it was assuming a more precise platform executing the
floating point operations. To work around this, this commit updates the
tests to run with reduced precision assertions when running under miri.
Otherwise the tests retain their default behavior.

mtreinish commented:
Hmm, I ran a quick benchmark and it is flagging a regression here

```
Benchmarks that have stayed the same:

| Change   | Before [6fbe13e3]    | After [3e4dae30]    | Ratio   | Benchmark (Parameter)                                                                                  |
|----------|----------------------|---------------------|---------|--------------------------------------------------------------------------------------------------------|
|          | 86.6±10ms            | 76.9±2ms            | ~0.89   | transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(1)                                |
|          | 0                    | 0                   | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('cx')                                          |
|          | 0                    | 0                   | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('cz')                                          |
|          | 0                    | 0                   | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('ecr')                                         |
|          | 566±10ms             | 619±3ms             | 1.09    | utility_scale.UtilityScaleBenchmarks.time_qv('cz')                                                     |
|          | 28.3±2ms             | 30.5±0.1ms          | 1.08    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cx')                           |
|          | 39.6±0.2ms           | 42.3±0.4ms          | 1.07    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(2)                               |
|          | 561±7ms              | 601±9ms             | 1.07    | utility_scale.UtilityScaleBenchmarks.time_qv('ecr')                                                    |
|          | 9.20±0.2ms           | 9.76±0.7ms          | 1.06    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(3)        |
|          | 89.5±1ms             | 94.5±1ms            | 1.06    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(3)                               |
|          | 87.8±3ms             | 92.7±2ms            | 1.06    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('ecr')                                        |
|          | 456±10ms             | 486±10ms            | 1.06    | utility_scale.UtilityScaleBenchmarks.time_qv('cx')                                                     |
|          | 65.3±2ms             | 68.8±0.4ms          | 1.05    | transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1)                   |
|          | 129±4ms              | 135±0.9ms           | 1.05    | transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2)                   |
|          | 21.9±0.3ms           | 23.0±0.7ms          | 1.05    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(1)                               |
|          | 27.7±0.2ms           | 29.3±2ms            | 1.05    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cz')                           |
|          | 333±5ms              | 344±2ms             | 1.04    | transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(3)                   |
|          | 636                  | 660                 | 1.04    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(2)                        |
|          | 102±2ms              | 106±2ms             | 1.04    | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('ecr')                                     |
|          | 605                  | 625                 | 1.03    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(3)                        |
|          | 30.1±1ms             | 31.0±1ms            | 1.03    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('ecr')                          |
|          | 191±9ms              | 198±4ms             | 1.03    | utility_scale.UtilityScaleBenchmarks.time_qaoa('cx')                                                   |
|          | 90.2±0.9ms           | 92.8±1ms            | 1.03    | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cx')                                      |
|          | 406±20ms             | 416±10ms            | 1.02    | utility_scale.UtilityScaleBenchmarks.time_qft('ecr')                                                   |
|          | 16.5±0.6ms           | 16.7±0.2ms          | 1.01    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(0)                          |
|          | 15.3±0.7ms           | 15.5±0.4ms          | 1.01    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(0)                               |
|          | 6.57±0.3ms           | 6.67±0.3ms          | 1.01    | utility_scale.UtilityScaleBenchmarks.time_bvlike('cz')                                                 |
|          | 8.51±0.3ms           | 8.64±0.5ms          | 1.01    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cx')                                        |
|          | 8.73±0.5ms           | 8.83±0.3ms          | 1.01    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cz')                                        |
|          | 241±5ms              | 243±5ms             | 1.01    | utility_scale.UtilityScaleBenchmarks.time_qaoa('ecr')                                                  |
|          | 296±8ms              | 299±4ms             | 1.01    | utility_scale.UtilityScaleBenchmarks.time_qft('cx')                                                    |
|          | 21.2±0.3ms           | 21.2±0.4ms          | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(1)                          |
|          | 13.9±0.4ms           | 13.9±0.7ms          | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(3)                          |
|          | 1429                 | 1429                | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(0)            |
|          | 1349                 | 1349                | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(1)            |
|          | 1352                 | 1352                | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(2)            |
|          | 1366                 | 1366                | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(3)            |
|          | 2705                 | 2705                | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(0)                   |
|          | 2005                 | 2005                | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(1)                   |
|          | 7                    | 7                   | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(2)                   |
|          | 7                    | 7                   | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(3)                   |
|          | 11117                | 11117               | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(0) |
|          | 5015                 | 5015                | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(1) |
|          | 16                   | 16                  | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(2) |
|          | 16                   | 16                  | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(3) |
|          | 1035                 | 1035                | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(0)                        |
|          | 762                  | 762                 | 1.00    | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(1)                        |
|          | 67.7±3ms             | 67.8±0.7ms          | 1.00    | utility_scale.UtilityScaleBenchmarks.time_bv_100('ecr')                                                |
|          | 845±50ms             | 846±40ms            | 1.00    | utility_scale.UtilityScaleBenchmarks.time_circSU2('cx')                                                |
|          | 86.5±5ms             | 86.3±1ms            | 1.00    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cz')                                         |
|          | 400                  | 400                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('cx')                                          |
|          | 400                  | 400                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('cz')                                          |
|          | 400                  | 400                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('ecr')                                         |
|          | 300                  | 300                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('cx')                                         |
|          | 300                  | 300                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('cz')                                         |
|          | 300                  | 300                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('ecr')                                        |
|          | 1714                 | 1714                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cx')                                            |
|          | 1727                 | 1727                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cz')                                            |
|          | 1727                 | 1727                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('ecr')                                           |
|          | 2012                 | 2012                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cx')                                             |
|          | 2053                 | 2053                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cz')                                             |
|          | 2053                 | 2053                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('ecr')                                            |
|          | 2463                 | 2463                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cx')                                              |
|          | 2463                 | 2463                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cz')                                              |
|          | 2463                 | 2463                | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('ecr')                                             |
|          | 432                  | 432                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cx')                               |
|          | 432                  | 432                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cz')                               |
|          | 432                  | 432                 | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('ecr')                              |
|          | 8.52±0.3ms           | 8.35±0.5ms          | 0.98    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(2)        |
|          | 6.91±0.1ms           | 6.75±0.1ms          | 0.98    | utility_scale.UtilityScaleBenchmarks.time_bvlike('cx')                                                 |
|          | 803±30ms             | 785±40ms            | 0.98    | utility_scale.UtilityScaleBenchmarks.time_circSU2('cz')                                                |
|          | 11.7±0.1ms           | 11.3±0.5ms          | 0.97    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(2)                          |
|          | 67.7±0.9ms           | 65.4±3ms            | 0.97    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(0)        |
|          | 8.70±0.4ms           | 8.42±0.5ms          | 0.97    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('ecr')                                       |
|          | 89.6±3ms             | 86.6±4ms            | 0.97    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cx')                                         |
|          | 75.4±0.5ms           | 72.6±3ms            | 0.96    | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(1)        |
|          | 832±20ms             | 794±30ms            | 0.95    | utility_scale.UtilityScaleBenchmarks.time_circSU2('ecr')                                               |
|          | 6.95±0.6ms           | 6.34±0.2ms          | 0.91    | utility_scale.UtilityScaleBenchmarks.time_bvlike('ecr')                                                |

Benchmarks that have got worse:

| Change   | Before [6fbe13e3]    | After [3e4dae30]    |   Ratio | Benchmark (Parameter)                                                                |
|----------|----------------------|---------------------|---------|--------------------------------------------------------------------------------------|
| +        | 255±6ms              | 385±7ms             |    1.51 | utility_scale.UtilityScaleBenchmarks.time_qaoa('cz')                                 |
| +        | 394±20ms             | 585±10ms            |    1.49 | utility_scale.UtilityScaleBenchmarks.time_qft('cz')                                  |
| +        | 68.6±3ms             | 88.5±0.7ms          |    1.29 | utility_scale.UtilityScaleBenchmarks.time_bv_100('cz')                               |
| +        | 101±0.7ms            | 131±3ms             |    1.29 | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cz')                    |
| +        | 64.5±0.4ms           | 80.0±0.5ms          |    1.24 | utility_scale.UtilityScaleBenchmarks.time_bv_100('cx')                               |
| +        | 34.3±0.4ms           | 38.2±0.3ms          |    1.11 | transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0) |
| +        | 86.0±3ms             | 94.6±0.9ms          |    1.1  | transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(0)              |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

If we don't have an alternative to make things in #14751 work on windows we don't really have a choice I guess. But I would like to know why #14751 only triggers the failure from C and we've not seen it from Python before. It feels like we're missing a piece of the puzzle if we're only failing from C.

raynelfss commented on .github/workflows/miri.yml:
```diff
@@ -14,15 +14,15 @@ jobs:
     name: Miri
     runs-on: ubuntu-latest
     env:
-      RUSTUP_TOOLCHAIN: nightly-2024-05-24
```
This is more of a question here but to update this here, what do we need to base upon?

mtreinish replied:
There is not a hard rule, we just need a version that's newer than the msrv and it needs to be fixed so a potential bug on a new nightly doesn't cause a spurious failure.

raynelfss commented on crates/quantum_info/src/sparse_pauli_op.rs:
```diff
@@ -1414,7 +1417,16 @@ mod tests {
                 .unwrap();
             let expected: DecomposeMinimal = paulis.into();
             let actual: DecomposeMinimal = decompose_dense_inner(arr.view(), 0.0).unwrap().into();
+            #[cfg(not(miri))]
             assert_eq!(actual, expected);
+            #[cfg(miri)]
+            {
+                assert!(actual.coeffs.abs_diff_eq(&expected.coeffs, 1e-8));
+                assert_eq!(actual.z, expected.z);
+                assert_eq!(actual.x, expected.x);
+                assert_eq!(actual.phases, expected.phases);
+                assert_eq!(actual.num_qubits, expected.num_qubits);
+            }
```
Why is `DecomposeMinimal`'s `PartialEq` implementation not miri safe?

mtreinish replied:
The details are in the commit message that made this change: https://github.com/Qiskit/qiskit/pull/14873/commits/3e4dae30caa19953852c593ef3ea207f4652d662 miri has a purposely low precision float implementation so the float comparisons need to have a larger tolerance to work with miri

raynelfss reviewed (commented):
My review here is mostly just questions about the changes in code that I didn't quite understand. I think everything looks good, and I can approve it after the merge conflict is fixed.

alexanderivrii commented:
The SVD in the current version of faer (0.19.4) is buggy: I have a small reproducer (on my arm-based mac laptop) that builds a specific 16x16 matrix where faer's SVD gives a completely wrong result. However, after rebasing on top of this PR, the problem goes away. Furthermore, by experimenting with #14797 (using faer's SVD instead of nalgebra's one), it seems that faer provides better numerical accuracy than nalgebra, in particular all of the python tests for QSD pass (again on my laptop) with the original error of `1e-8` (including the larger tests: 2-controlled 4 qubit-unitaries and 1-controlled 5-qubit unitaries). Hence I eager to see this PR merged.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into bump-faer-and-msrv

mtreinish commented:
I think this is ready to move forward. I ran the benchmarks again and it is still showing a regression on some of the utility scale benchmarks but also a smaller improvement on some other benchmarks now. But given the correctness issues @alexanderivrii reported this fixes and also mitigating the segfault in #14751 we should move forward here and we can see if there are other mitigations for the performance regression asv is showing.

raynelfss reviewed (commented):
This looks about ready to merge. You do need to upgrade the `wheels-build.yml` workflow since it installs 1.79 by default.

mtreinish commented:
> This looks about ready to merge. You do need to upgrade the `wheels-build.yml` workflow since it installs 1.79 by default.

It actually doesn't (or if it does that's a mistake and we shouldn't). By default we build all release wheels with the latest stable version of rust. The exception is for platforms that had build failures with the latest stable, which use the `install_rust_msrv.sh` script to install rust inside the container at runtime before the builds begin. IIRC the only one that fell into this category was the aarch64 linux wheels. But, that is a good reminder to check that it actually builds with PGO and LTO on rust 1.85 because there were issues when we promoted that to tier 1 on newer rust versions. Lets add the tag to run the wheel builds.

mtreinish committed:
Try using latest stable with aarch64 wheel builds

mtreinish commented:
We're able to use the latest stable for aarch64 linux builds now so I made that change in: https://github.com/Qiskit/qiskit/pull/14873/commits/cae731230cce195cc9c08fa1bacffc4881c47133

raynelfss commented on .github/workflows/wheels-build.yml:
```diff
@@ -101,8 +101,8 @@ jobs:
+              RUST_TOOLCHAIN=stable
           else
               INSTALL_RUST_PATH=tools/install_rust.sh
               RUST_TOOLCHAIN=stable
```
Since both paths lead to the same outcome, should we remove these conditions?

mtreinish replied:
I left it in place just in case we have issues with aarch64 in the future. But I can remove it if you'd prefer

raynelfss replied:
You can leave it as long as you add a comment about why the condition is there. But at that point it might just be better to remove it and add it if necessary. Up to you.

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/14873/commits/41b149176fdf1215f1aceb4d63253e84c177bb77

raynelfss reviewed (commented):
One more quick question here :eyes:

mtreinish committed:
Remove switchable rust compiler version as it is no longer needed

raynelfss reviewed (approved):
LGTM
